docs/fakerclass.rst

```diff
@@ -0,0 +1,289 @@
+we did for ``seed_instance``, but the rest like `add_provider` and the `random` getter and setter
+are more dependent on specific use cases or are potentially dangerous.
+
+In those cases, it is better for users to create their own subclass with their implementation or to
+directly call those methods from the internal ``Generator`` objects themselves. Multiple locale mode
+will be discussed in more detail in its `dedicated section`_.
+
+Proxy Class Attribute Name Resolution
+-------------------------------------
+
+The proxy class has a fairly involved attribute name resolution behavior that runs in this order:
+
+1. If the attribute name is `seed`, raise a TypeError. This prevents the class method `seed`
+   from being called from an instance.
+2. If #1 does not apply, check if the attribute name matches an attribute present in the proxy
+   class instance. If there is one, return the matching attribute.
+3. If #2 failed, check if the instance is in single locale mode. If yes, proxy the call to the
+   sole internal ``Generator`` object, and attempt to return a matching attribute.
+4. If #3 does not apply, the instance is henceforth known to be in multiple locale mode. Proceed
+   by checking if the attribute name matches a ``Generator`` attribute. If it does, raise a
+   NotImplementedError.
+5. If #4 does not apply, check if the attribute name matches a cache pattern regex. If it does not,
+   raise an AttributeError, since it should already have been handled by #2 if one does exist.
+6. If everything else has failed or does not apply, assume that the attribute name might be
+   referring to a provider method and perform factory/generator selection, and proxy the call
+   to the selected ``Generator`` object.
+
+Factory/generator selection will be discussed in more detail under multiple locale mode's
+`dedicated section`_.
+
+Locale Normalization
+--------------------
+
+Depending on the ``locale`` value passed, a new ``Faker`` instance will either operate in single
+locale mode or multiple locale mode. The value of ``locale`` can be one of the following:
+
+1. Any empty value like `None` (automatically defaults to en_US)
+2. A valid locale string, underscored or hyphenated
+3. A list, tuple, or set with valid locale strings, underscored or hyphenated
+4. An OrderedDict with key-value pairs of valid locale strings (underscored or
+   hyphenated) and weights
+
+The first two are options already expected by old ``Faker``, so it is pretty much the same for new
+``Faker``. Using any of those two options will always result in a new ``Faker`` instance that is
+in single locale mode. In that mode, there is really no need to retrieve a reference to the
+internal ``Generator`` object because of the 1:1 proxying behavior discussed earlier.
+
+The potential pitfalls lie in multiple locale mode and when there is a need to access the internal
+``Generator`` objects individually. Since locale strings can be written underscored (en_US) or
```

Comment:
I'd liuke to have all locale strings formatted as code:

```rst
``Generator`` objects individually. Since locale strings can be written underscored (``en_US``) or
hyphenated (``en-US``)
```